runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java

```diff
@@ -82,43 +104,27 @@ public void cleanup() throws Exception {
       CommittedBundle<InputT> inputBundle)
       throws Exception {
     final ProcessElements<InputT, OutputT, RestrictionT, TrackerT> transform =
-        application.getTransform();
-
-    ProcessFn<InputT, OutputT, RestrictionT, TrackerT> processFn =
-        transform.newProcessFn(transform.getFn());
-
-    DoFnLifecycleManager fnManager = DoFnLifecycleManager.of(processFn);
-    processFn =
-        ((ProcessFn<InputT, OutputT, RestrictionT, TrackerT>)
-            fnManager.<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>get());
-
-    String stepName = evaluationContext.getStepName(application);
-    final DirectExecutionContext.DirectStepContext stepContext =
-        evaluationContext
-            .getExecutionContext(application, inputBundle.getKey())
-            .getStepContext(stepName);
-
-    final ParDoEvaluator<KeyedWorkItem<String, KV<InputT, RestrictionT>>>
-        parDoEvaluator =
-            delegateFactory.createParDoEvaluator(
-                application,
-                inputBundle.getKey(),
-                (PCollection<KeyedWorkItem<String, KV<InputT, RestrictionT>>>)
-                    inputBundle.getPCollection(),
-                transform.getSideInputs(),
-                transform.getMainOutputTag(),
-                transform.getAdditionalOutputTags().getAll(),
-                stepContext,
-                processFn,
-                fnManager);
-
-    processFn.setStateInternalsFactory(key -> (StateInternals) stepContext.stateInternals());
-
-    processFn.setTimerInternalsFactory(key -> stepContext.timerInternals());
+      application.getTransform();
+    final DoFnLifecycleManagerRemovingTransformEvaluator
+      <KeyedWorkItem<String, KV<InputT, RestrictionT>>> evaluator =
+      delegateFactory.createEvaluator(
+        (AppliedPTransform) application,
+        (PCollection<KeyedWorkItem<String, KV<InputT, RestrictionT>>>) inputBundle.getPCollection(),
+        inputBundle.getKey(),
+        application.getTransform().getSideInputs(),
+        application.getTransform().getMainOutputTag(),
+        application.getTransform().getAdditionalOutputTags().getAll());
+
+    final ProcessFn<InputT, OutputT, RestrictionT, TrackerT> processFn =
+      (ProcessFn<InputT, OutputT, RestrictionT, TrackerT>)
+        ProcessFnRunner.class.cast(evaluator.underlying.fnRunner).getFn();
+    processFn.setStateInternalsFactory(key -> evaluator.underlying.stepContext.stateInternals());
+
+    processFn.setTimerInternalsFactory(key -> evaluator.underlying.stepContext.timerInternals());
 
     OutputWindowedValue<OutputT> outputWindowedValue =
         new OutputWindowedValue<OutputT>() {
-          private final OutputManager outputManager = parDoEvaluator.getOutputManager();
+          private final OutputManager outputManager = evaluator.underlying.getOutputManager();
```

Comment:
What motivates this change in access pattern (and the other changes to `delegateFactory` fields)? It seems sketchy.

Reply:
i found this compromise better to reuse the existing logic of the DoFn than creating a new internal API to access the data (kind of unwrapper). Do you prefer anemic getters?

Reply:
It breaks basic abstraction boundaries - we don't get to see the internal details of how the nested evaluator manages its contexts, and we're responsible for managing our own context (DoFnLifecycleManagerRemovingTransformEvaluator has a contract that we depend upon, and implementation we do not).

Reply:
Will move to getters but since it is all about internals of the direct runner it should be more than fine. Te intent being to fully delegate we must expose the underlying anyway - otherwise we duplicate code which would be worse IMHO.